File: java21/src/main/java/io/github/sps23/designpatterns/facade/OrderFulfillmentFacade.java
package io.github.sps23.designpatterns.facade;

import java.math.BigDecimal;
import java.util.Objects;

public final class OrderFulfillmentFacade {
    private final InventoryGateway inventoryGateway;
    private final PaymentGateway paymentGateway;
    private final ShippingGateway shippingGateway;
    private final NotificationGateway notificationGateway;

    public OrderFulfillmentFacade(InventoryGateway inventoryGateway, PaymentGateway paymentGateway,
            ShippingGateway shippingGateway, NotificationGateway notificationGateway) {
        this.inventoryGateway = Objects.requireNonNull(inventoryGateway,
                "Inventory gateway must not be null");
        this.paymentGateway = Objects.requireNonNull(paymentGateway,
                "Payment gateway must not be null");
        this.shippingGateway = Objects.requireNonNull(shippingGateway,
                "Shipping gateway must not be null");
        this.notificationGateway = Objects.requireNonNull(notificationGateway,
                "Notification gateway must not be null");
    }

    public FulfillmentResult placeOrder(OrderRequest request) {
        Objects.requireNonNull(request, "Order request must not be null");

        if (!inventoryGateway.hasStock(request.sku(), request.quantity())) {
            return FulfillmentResult.failure("Inventory unavailable for SKU " + request.sku());
        }

        if (!paymentGateway.charge(request.customerId(), request.amount())) {
            return FulfillmentResult.failure("Payment failed for customer " + request.customerId());
        }

        String trackingId = shippingGateway.scheduleShipment(request.customerId(),
                request.shippingAddress());
        notificationGateway.sendConfirmation(request.customerId(), request.sku(), trackingId);

        return FulfillmentResult.success("Order placed successfully", trackingId);
    }

    public interface InventoryGateway {
        boolean hasStock(String sku, int quantity);
    }

    public interface PaymentGateway {
        boolean charge(String customerId, BigDecimal amount);
    }

    public interface ShippingGateway {
        String scheduleShipment(String customerId, String shippingAddress);
    }

    public interface NotificationGateway {
        void sendConfirmation(String customerId, String sku, String trackingId);
    }

    public record OrderRequest(String customerId, String sku, int quantity, BigDecimal amount,
            String shippingAddress) {

        public OrderRequest {
            Objects.requireNonNull(customerId, "Customer ID must not be null");
            Objects.requireNonNull(sku, "SKU must not be null");
            Objects.requireNonNull(amount, "Amount must not be null");
            Objects.requireNonNull(shippingAddress, "Shipping address must not be null");
            if (quantity <= 0) {
                throw new IllegalArgumentException("Quantity must be positive");
            }
            if (amount.compareTo(BigDecimal.ZERO) <= 0) {
                throw new IllegalArgumentException("Amount must be greater than zero");
            }
        }
    }

    public record FulfillmentResult(boolean success, String message, String trackingId) {
        public static FulfillmentResult success(String message, String trackingId) {
            return new FulfillmentResult(true, message, trackingId);
        }

        public static FulfillmentResult failure(String message) {
            return new FulfillmentResult(false, message, null);
        }
    }

    public static final class InventoryService implements InventoryGateway {
        @Override
        public boolean hasStock(String sku, int quantity) {
            return "SKU-42".equals(sku) && quantity <= 5;
        }
    }

    public static final class PaymentService implements PaymentGateway {
        @Override
        public boolean charge(String customerId, BigDecimal amount) {
            return !"blocked-customer".equals(customerId) && amount.compareTo(BigDecimal.ZERO) > 0;
        }
    }

    public static final class ShippingService implements ShippingGateway {
        @Override
        public String scheduleShipment(String customerId, String shippingAddress) {
            if (shippingAddress == null || shippingAddress.isBlank()) {
                throw new IllegalArgumentException("Shipping address must not be blank");
            }
            return "TRACK-" + customerId.toUpperCase() + "-" + shippingAddress.hashCode();
        }
    }

    public static final class NotificationService implements NotificationGateway {
        @Override
        public void sendConfirmation(String customerId, String sku, String trackingId) {
            System.out.println("Sending confirmation to " + customerId + " for " + sku
                    + " with tracking ID " + trackingId);
        }
    }
}

File: scala3/src/main/scala/io/github/sps23/designpatterns/facade/OrderFulfillmentFacade.scala
package io.github.sps23.designpatterns.facade

import java.math.BigDecimal

class OrderFulfillmentFacade(
    inventoryGateway: OrderFulfillmentFacade.InventoryGateway,
    paymentGateway: OrderFulfillmentFacade.PaymentGateway,
    shippingGateway: OrderFulfillmentFacade.ShippingGateway,
    notificationGateway: OrderFulfillmentFacade.NotificationGateway
):
  def placeOrder(
      request: OrderFulfillmentFacade.OrderRequest
  ): OrderFulfillmentFacade.FulfillmentResult =
    require(request != null, "Order request must not be null")

    if !inventoryGateway.hasStock(request.sku, request.quantity) then
      return OrderFulfillmentFacade.FulfillmentResult.failure(
        s"Inventory unavailable for SKU ${request.sku}"
      )

    if !paymentGateway.charge(request.customerId, request.amount) then
      return OrderFulfillmentFacade.FulfillmentResult.failure(
        s"Payment failed for customer ${request.customerId}"
      )

    val trackingId = shippingGateway.scheduleShipment(request.customerId, request.shippingAddress)
    notificationGateway.sendConfirmation(request.customerId, request.sku, trackingId)

    OrderFulfillmentFacade.FulfillmentResult.success(
      "Order placed successfully",
      trackingId
    )

object OrderFulfillmentFacade:
  trait InventoryGateway:
    def hasStock(sku: String, quantity: Int): Boolean

  trait PaymentGateway:
    def charge(customerId: String, amount: BigDecimal): Boolean

  trait ShippingGateway:
    def scheduleShipment(customerId: String, shippingAddress: String): String

  trait NotificationGateway:
    def sendConfirmation(customerId: String, sku: String, trackingId: String): Unit

  case class OrderRequest(
      customerId: String,
      sku: String,
      quantity: Int,
      amount: BigDecimal,
      shippingAddress: String
  ):
    require(customerId != null && customerId.nonEmpty, "Customer ID must not be blank")
    require(sku != null && sku.nonEmpty, "SKU must not be blank")
    require(quantity > 0, "Quantity must be positive")
    require(
      amount != null && amount.compareTo(BigDecimal.ZERO) > 0,
      "Amount must be greater than zero"
    )
    require(
      shippingAddress != null && shippingAddress.nonEmpty,
      "Shipping address must not be blank"
    )

  case class FulfillmentResult(success: Boolean, message: String, trackingId: Option[String])

  object FulfillmentResult:
    def success(message: String, trackingId: String): FulfillmentResult =
      FulfillmentResult(true, message, Some(trackingId))

    def failure(message: String): FulfillmentResult =
      FulfillmentResult(false, message, None)

  class InventoryService extends InventoryGateway:
    override def hasStock(sku: String, quantity: Int): Boolean = sku == "SKU-42" && quantity <= 5

  class PaymentService extends PaymentGateway:
    override def charge(customerId: String, amount: BigDecimal): Boolean =
      customerId != "blocked-customer" && amount.compareTo(BigDecimal.ZERO) > 0

  class ShippingService extends ShippingGateway:
    override def scheduleShipment(customerId: String, shippingAddress: String): String =
      s"TRACK-${customerId.toUpperCase}-${shippingAddress.hashCode()}"

  class NotificationService extends NotificationGateway:
    override def sendConfirmation(customerId: String, sku: String, trackingId: String): Unit =
      println(s"Sending confirmation to $customerId for $sku with tracking ID $trackingId")
